### PR TITLE
hot restart: clear initializing in abnormal exit case

### PR DESCRIPTION
**Pull request by mattklein123**
See code comment for more info.

Fixes https://github.com/lyft/envoy/issues/726

### CODE REVIEW

**Comment by mattklein123** [comment]
@kyessenov can you please smoke test this?

**Comment by kyessenov** [comment]
I still haven't done the cluster rename so can't take the master of Envoy and make it run in Istio easily. I'll see if I can just simulate the failure with a dumb config.

**Comment by kyessenov** [comment]
I'm thinking that we should do same thing as your hot restarter and tear down everything on abnormal exit. I still see assertion errors in hot restart code sometimes when using TLS certs, although the next restart succeeds and the proxies recover.

**Comment by mattklein123** [comment]
> I still see assertion errors in hot restart code sometimes when using TLS certs, although the next restart succeeds and the proxies recover.

You mean even with this change? Can you be more specific about the production asserts you are seeing?

**Comment by kyessenov** [comment]
We only test debug envoy images. I'll try to recover a log, but from the logs it looked like that if config has SSL context on listeners and RDS does not complete in time before hot restart is invoked (we set aggressive hot restart frequency < 1s), the proxy crashes with an abnormal exit.

**Comment by mattklein123** [comment]
OK well LMK if you want this fix or not. If not we can close the original issue.

File: source/exe/hot_restart.cc [hunk 1]
```diff
@@ -103,6 +103,19 @@ HotRestartImpl::HotRestartImpl(Options& options)
   UNREFERENCED_PARAMETER(rc);
 }
 
+HotRestartImpl::~HotRestartImpl() {
+  if (clear_initialized_on_exit_) {
+    // In this case, we are exiting before we cleared initializing ourself. This can happen if for
+    // example we try to load an invalid config. We can't universally clear initializing in shared
+    // memory since we might be gracefully shutting down during normal hot restart processing. Also,
+    // it should be noted that the general recommendation is to us something like restarter.py which
```
Comment (ccaraman) [inline source/exe/hot_restart.cc]: typo: use

File: source/exe/hot_restart.cc [hunk 2]
```diff
@@ -103,6 +103,19 @@ HotRestartImpl::HotRestartImpl(Options& options)
...
+    // tears everything down in the case of any abnormal exit. However this is a basic safeguard
+    // tha can work if a new process has a config error when starting up before it actually begins
```
Comment (ccaraman) [inline source/exe/hot_restart.cc]: typo:tha ->that

**Comment by mattklein123** [comment]
Per discussion with @kyessenov closing this since a safer solution for runners is to kill all Envoys in the case of abnormal exit.

**Pull request closed without merging**